internal/expr/bindinputs.go

```diff
@@ -41,66 +41,61 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
 	argTypeUsed := map[reflect.Type]bool{}
 	inputCount := 0
 	outputCount := 0
-	sqlStr := bytes.Buffer{}
+	var sqlBuilder sqlBuilder
 	for _, te := range *tbe {
 		switch te := te.(type) {
 		case *typedInputExpr:
-			vals, err := te.input.LocateParams(typeToValue)
+			vals, omit, err := te.input.LocateParams(typeToValue)
 			if err != nil {
 				return nil, err
 			}
+
+			if omit {
+				return nil, omitEmptyInputError(te.input.Desc())
+			}
 			argTypeUsed[te.input.ArgType()] = true
-			for i, val := range vals {
-				if i != 0 {
-					sqlStr.WriteString(", ")
-				}
+			sqlBuilder.writeInput(inputCount, len(vals))
+			for _, val := range vals {
 				namedInput := sql.Named("sqlair_"+strconv.Itoa(inputCount), val.Interface())
 				params = append(params, namedInput)
-				sqlStr.WriteString("@sqlair_" + strconv.Itoa(inputCount))
 				inputCount++
 			}
 		case *typedInsertExpr:
-			// Write out the columns.
-			sqlStr.WriteString("(")
-			for i, col := range te.columns {
-				if i != 0 {
-					sqlStr.WriteString(", ")
-				}
-				sqlStr.WriteString(col)
-			}
-			sqlStr.WriteString(") VALUES (")
-			// Write out the values.
-			for i, input := range te.inputs {
-				if i != 0 {
-					sqlStr.WriteString(", ")
-				}
-				vals, err := input.LocateParams(typeToValue)
+			var columns []string
+			for _, ic := range te.insertColumns {
+				vals, omit, err := ic.input.LocateParams(typeToValue)
 				if err != nil {
 					return nil, err
 				}
-				argTypeUsed[input.ArgType()] = true
-				for j, val := range vals {
-					if j != 0 {
-						sqlStr.WriteString(", ")
+				if len(vals) > 1 {
+					// Only slices return multiple values and they are not
+					// allowed in insert expressions.
+					return nil, fmt.Errorf("internal error: unexpected values")
```

Comment:
The error might be more descriptive if the text is along the lines of the comment above.

Reply:
Changed.